### ISSUE TITLE
carbonserver + grafana issue?

### ISSUE THREAD

**Issue by rlex** [go-graphite/go-carbon#130]
If i understood correctly, carbonserver is replacement to existing python-based graphite-web so we can feed whisper data directly to, for example, grafana?

I tried installing RC and using it with grafana, however

1) on adding grafana says "unknown error", and in log i see it's trying to access /metrics/find URL which throws 404 by go-carbon, ie:
```
t=2016-11-28T02:32:24+0300 lvl=info msg="Request Completed" logger=context userId=2 orgId=1 uname=lex method=GET path=/metrics/find/ status=404 remote_addr=127.0.0.1 time_ms=5ns size=17945
```

2) Tried accessing /metrics/find with cURL and getting 301

Looks like grafana expects /metrics/find/ in URL (with trailing slash), however it throws 301 on such request, which grafana can't process:
```
❯ curl -v http://localhost:8080/metrics/find
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /metrics/find HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 301 Moved Permanently
< Location: /metrics/find/
< Vary: Accept-Encoding
< Date: Mon, 28 Nov 2016 14:47:23 GMT
< Content-Length: 49
< Content-Type: text/html; charset=utf-8
<
<a href="/metrics/find/">Moved Permanently</a>.

* Connection #0 to host localhost left intact
```

And this is with /metrics/find/:
```
❯ curl -v "http://localhost:8080/metrics/find/?query=carbon.*"
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /metrics/find/?query=carbon.* HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 400 Bad Request
< Content-Type: text/plain; charset=utf-8
< Vary: Accept-Encoding
< X-Content-Type-Options: nosniff
< Date: Mon, 28 Nov 2016 14:51:09 GMT
< Content-Length: 33
<
Bad request (unsupported format)
* Connection #0 to host localhost left intact
```

I'm not sure if this is grafana or go-carbon bug

**Comment by Civil** [comment]
It's not a complete replacement for graphite-web. It can only replace local graphite-web for CLUSTER_SERVERS needs.

If you want to have a replacement for graphite-web (also subset that's enough for grafana) you can get https://github.com/dgryski/carbonapi and point it to go-carbon's carbonserver.

So the setup you want should look like that:

grafana -> carbonapi -> go-carbon

**Comment by rlex** [comment]
Thanks, seems like i misunderstood purpose of carbonserver.

P.S. no plans to make go-carbon completely self-sufficient, including an API for external tools?

**Comment by piotr1212** [comment]
That might make sense for smaller single server setups. I'm assuming most go-carbon users run larger installations and need a more distributed architecture with loadbalancers, relays, multiple carbon backends, zypper (or multiple layers of graphite-web), carbon-api, grafana. Anyway, I'm not involved in go-carbon so I have no idea what the plans actually are, just my thoughts.

**Comment by deniszh** [comment]
I also do not know plans of the author regarding that, but IMO running another binary is not a rocket science. Carbonapi has quite good development speed, so, including that into gocarbon is not an option. You need to run grafana as separate binary anyway...

**Comment by lomik** [comment]
I think that is sufficient to maintain the compatibility of go-carbon with carbonapi. Not sure what they are currently working together without carbonzipper

**Comment by atlantageek** [comment]
Is there any guidance on how to get this working.  Im trying to build an evaluation server for our product. I have go-carbon and grafana installed but it was not clear how to get carbonapi running.  looks like it requires carbonzipper? Is this true. I was hoping to build a single server so I could compare it to influxdb.

**Comment by deniszh** [comment]
Yes, carbonzipper is required, as [carbonapi repo](https://github.com/dgryski/carbonapi) states - "To use this, you must have a carbonzipper install, which in turn requires that your carbon stores are running carbonserver". You need to point carbonapi to carbonzipper, and carbonzipper to carbonserver. It's little bit harder than simple influxdb setup, but it's much more flexible and scalable IMO.

**Comment by rlex** [comment]
@atlantageek you don't need carbonzipper with go-carbon (at least for current rc version).

So, if you want to evaluate some stuff, install -rc go-carbon, carbonapi (if you're on debian/ubuntu you can use my packaging repo: https://github.com/rlex/carbonapi-deb)
My packaging repo includes systemd init as well, but if you want to build & start it manually, you can go with 

```
./carbonapi -z=http://127.0.0.1:8080 -p 8082
```

where http://127.0.0.1:8080 is carbonserver address in go-carbon (available only in -rc version ATM). It's in this section of go-carbon.conf config:
```
[carbonserver]
listen = "127.0.0.1:8080"
enabled = true
buckets = 10
max-globs = 100
metrics-as-counters = false
read-timeout = "1m0s"
write-timeout = "1m0s"
scan-frequency = "5m0s"
```

UPD: and point grafana instance to carbonapi, in case of my example that will be port 8082.

**Comment by Civil** [comment]
Actually carbonzipper is not required for grafana to work. You can point carbonapi directly to carbonserver and everything except for a custom /info handler will work.

**Comment by StephenPCG** [comment]
I'm trying to setup a monitor stack for a large cluster (expected metrics rate > 10,000,000 per minute).
I happen to come across this issue, and I would like to explain the architecture of my current setup, for those who are interested. If there are mistakes, please kindly point out.

The "official" graphite stack looks like this:

```
                     |-----------------|
                     | carbon-relay.py |
                     |-----------------|
                              |
        |---------------------+---------------------|
|-----------------|  |-----------------|  |-----------------|
| carbon-cache.py |  | carbon-cache.py |  | carbon-cache.py |
|  graphite-web   |  |  graphite-web   |  |  graphite-web   |
|-----------------|  |-----------------|  |-----------------|
        |---------------------+---------------------|
                              |
                     |-----------------|
                     |  graphite-web   |
                     |     grafana     |
                     |-----------------|
```

If you have even larger cluster, you can have multiple ``carbon-relay.py`` instances, as long as they use the same relay rule.

Every ``carbon-cache.py`` instance comes along with a ``graphite-web``, ``graphite-web`` read metrics from ``carbon-cache.py``'s in-memory cache as well as local whisper files. There is another "higher level" ``graphite-web`` instance, which does not read any data from carbon-cache or local whisper files, it queries metrics from all configured upstreams, via ``CLUSTER_SERVERS`` option, and merge them. Finally, it provides render api's for grafana or other dashboards.

As you can see, all official graphite components are written in python, them have bad performance. Luckly, almost all components have third party re-implementation in c or go.

[carbon-relay-ng](https://github.com/graphite-ng/carbon-relay-ng) and [carbon-c-relay](https://github.com/grobian/carbon-c-relay) are replacement for ``carbon-relay.py``, implemented in go and c. I'm currently using ``carbon-relay-ng``. [go-carbon](https://github.com/lomik/go-carbon) is replacement for ``carbon-cache.py``. [carbonapi](https://github.com/dgryski/carbonapi) is replacement for ``graphite-web`` which is ready to serve data to grafana.

``carbonlink`` provides the same function as ``carbon-cache.py``'s in-memory cache. So you can have the following arch:

```
                     |-----------------|
                     | carbon-relay-ng |
                     |-----------------|
                              |
        |---------------------+---------------------|
|-----------------|  |-----------------|  |-----------------|
|    go-carbon    |  |    go-carbon    |  |    go-carbon    |
|  (carbonlink)   |  |  (carbonlink)   |  |  (carbonlink)   |
|  graphite-web   |  |  graphite-web   |  |  graphite-web   |
|-----------------|  |-----------------|  |-----------------|
        |---------------------+---------------------|
                              |
                     |-----------------|
                     |  graphite-web   |
                     |     grafana     |
                     |-----------------|
```

You may note that, the graphite-web along with carbon-cache need not to be a full functional graphite-web, it only need a few api's to provide metric data to the higher level graphite-web instance, carbonserver does it.

```
                     |-----------------|
                     | carbon-relay-ng |
                     |-----------------|
                              |
        |---------------------+---------------------|
|-----------------|  |-----------------|  |-----------------|
|    go-carbon    |  |    go-carbon    |  |    go-carbon    |
| (carbonserver)  |  | (carbonserver)  |  | (carbonserver)  |
|-----------------|  |-----------------|  |-----------------|
        |---------------------+---------------------|
                              |
                     |-----------------|
                     |  graphite-web   |
                     |     grafana     |
                     |-----------------|
```

The next step, graphite-web is implemented in python, and it renders slow when metric volume grows, you may like to replace it with carbonapi. If you have just one go-carbon instance, it's ok to directly replace graphite-web with carbonapi. However, carbonapi is only capable to query data from one upstream, so, if you have multiple go-carbon instances, you should introduce carbonzipper, which will fetch metric data from multiple upstreams and merge them. carbonapi, in turn, querys data from carbonzipper.

Here is the final architecture:

```
                     |-----------------|
                     | carbon-relay-ng |
                     |-----------------|
                              |
        |---------------------+---------------------|
|-----------------|  |-----------------|  |-----------------|
|    go-carbon    |  |    go-carbon    |  |    go-carbon    |
| (carbonserver)  |  | (carbonserver)  |  | (carbonserver)  |
|-----------------|  |-----------------|  |-----------------|
        |---------------------+---------------------|
                              |
                     |-----------------|
                     |  carbonzipper   |
                     |    carbonapi    |
                     |     grafana     |
                     |-----------------|
```

Every thing except grafana is implemented in go, which makes OPS's life much easier. It's nightmare to deploy and manage python projects. Grafana is much more a frontend project, it manages dashboards.

**Comment by Civil** [comment]
We are using something similar to your drawing (now also with go-carbon instead of carbon-cache). I've done a talk about our architecture some time ago, you can have a look at slides:
http://events.linuxfoundation.org/sites/events/files/slides/booking-graphite-atscale-linuxconeur2k16.pdf

**Comment by deniszh** [comment]
Your final architecture is fine, should work flawlessly. Grafana's backend is written in Go too btw.

**Comment by StephenPCG** [comment]
@Civil Thanks for your slides! It's really valuable for me. We're at the very beginning of setting this monitoring stack, I just have the whole stack running, the next step is stress testing and disaster exercise, to prove this stack is suitable for our future environment. Your slides gives me a lot of reference and confidence.

**Comment by ghost** [comment]
@StephenPCG Did you implement the stack as you described it within this ticket? If yes, are there any problems / bottlenecks?

**Comment by StephenPCG** [comment]
@alexanderlauster I didn't setup the exact stack, but with a simplified stack ``carbon-relay-ng -> (go-carbon+graphite-web)x6 -> graphite-web -> grafana``, this setup was just a proof of concept, and for a simplified stress testing, to find out the relationship among metric volume, event volume and disk iops.

We're going to setup the stack I described in this month, and run a more detailed stress test. 

Is there any problems? Since I haven't had this stack running in a production environment, I haven't came across any problems yet. But from some slides and videos found on the Internet, cluster re-balancing may be a complex work and have great burden.

Is there any bottlenecks? I think you will probably reach disk io limit before cpu and memory, if you are not using SSD RAID. In my stress test setup, I had 6 magnet disks, no RAID setup, running one go-carbon instance on each. With a small event rate (less then 3k metric points per second), the disk iops was the same as event rate. go-carbon does not merge metric points, it issues a write operation for each  point, and relies on underlying filesystem to merge write operations. When events rate kept growing up, disk iops stay at 3k (which is magnet disk limit), and filesystem started to merge write operations. ``dirty_ratio``, ``dirty_background_ratio``, ``dirty_expire_centisecs`` affect the filesystem cache, If you set a too small expire centisecs, or have to small dirty_*_ratio, the filesystem will not hold too many cache, and filesystem write won't catch up the incoming events. If you have expire centisecs set to 1 minute and large enough dirty_*_ratio for example, filesystem can cache and merge all write operations for one metric coming in in one minute, if the metric have 10s retention, it means 6 write operations can be merged into 1. But with larger expire centisecs and dirty ratio, you will loose more data on hardware failure. There is another interesting thing, "new metric rate". When a new metric received, go-carbon will create a new wsp file for that metric. The file size depends on retention setup. If file size is very large and new metric rate is very high, you may easily reach disk throughput limit.  In my stress test setup, a wsp file is about 800KB, on stress test startup, it created about 750K metrics, which corresponds to about 600GB wsp files, apparently 6 magnet disks can not accept so large data in a short time, go-carbon and carbon-relay-ng's cache quickly filled up and started to drop incoming metrics.

In short, you may probably be interested on these aspects when inspecting stack performance:
* event rate (how many metric points every second)
* metric volume (how many distinct metric series there, the more metric series, the more cache required)
* new metric rate (how many new distinct metric series created every second)

What I didn't test in the previous stress test? I didn't test metric rewriting and aggregation, these two tasks may be CPU intensive. I will do these tests this month. I didn't try re-balancing a cluster (add or remove go-carbon instances), I will try it later.

Hope these information helps.
